[dynamic shapes] unbacked-safe slicing

pianpwk opened the pull request:
Summary:
Generates new unbacked symbols for slice output size & storage offset, when appropriate semantics are unclear. Teaches inductor to codegen the slice with flexible semantics.


Test Plan:
contbuild & OSS CI, see https://hud.pytorch.org/commit/pytorch/pytorch/56218d85e2da09d9ede3809718ec989c2151632c

Rollback Plan:

Differential Revision: D80948073




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

laithsakka commented on torch/_inductor/lowering.py:
```diff
@@ -1172,9 +1172,110 @@ def permute(x, dims):
 def slice_(x, dim=0, start=0, end=2**63, step=1, clamp=True):
+    """
+    Lowers a slice call, creating ExternKernels for the output size & storage offset symbols,
+    if the indices are unbacked and appropriate semantics aren't known.
```
nit: semantics are the same for known and unknown. 

maybe:
creating ExternKernels to compute the output size & storage offset symbols dynamically if they  cant be determined at compile time because indices are unbacked.

laithsakka commented on torch/_inductor/lowering.py:
```diff
@@ -1172,9 +1172,110 @@ def permute(x, dims):
+                sym_storage = sym
+
+    if V.graph.current_node is None or not clamp or (sym_size is None and sym_storage is None):
+        return TensorBox(
```
// This path is taken if no unbacked symbols has been allocated meaning that we know how to determine things statically and no need for dynamic computations.

laithsakka commented on torch/_inductor/lowering.py:
```diff
@@ -1172,9 +1172,110 @@ def permute(x, dims):
+    except TypeError:
+        pass
+
+    # try to avoid dynamic (unbacked) slice
```
this does not seem the appropriate line for this comment?

laithsakka reviewed (approved):
approved but please address my comments before landing

pianpwk committed:
[dynamic shapes] unbacked-safe slicing (reland #157944) (#161414)

Summary:

Generates new unbacked symbols for slice output size & storage offset, when appropriate semantics are unclear. Teaches inductor to codegen the slice with flexible semantics.


Test Plan: contbuild & OSS CI, see https://hud.pytorch.org/commit/pytorch/pytorch/56218d85e2da09d9ede3809718ec989c2151632c

Reviewed By: laithsakka

Differential Revision: D80948073
